### ISSUE TITLE
Develop camera's metadata protocol

### ISSUE THREAD

**Issue by victor-wildlife** [wildlifeai/wildlife-watcher-mobile-app#54]
The images captured by the device should have "easy to recover" metadata (e.g. date, location, project info...).

There are different options:
1 - have a "metadata log" file that keeps track of the photos taken and their respective metadata,
2 - embed the metadata information in the videos/images directly (e.g. as EXIF data),
3 - a combination of both to reduce potential repetition and save on the size of images/video (even though EXIF data would be only a few KB (<1% of total image/video size).

My preferred option is #2, adding the information as EXIF reduces the chances of having photos/videos and not knowing where/when they come from. I am afraid this will require a bit of code engineering from our end though so I would love to know @Tobyntobyn and @acutetech thoughts on it.

The other part I would like us to think about is where in the camera should we store general metadata (e.g. device firmware, deployment location, specific settings...) and the format. Should we store it on the SD card or the internal memory of the grove vision/WW130? The camera will need to read it and add the information as EXIF to each picture/movie.
What would be the best format a .json, .csv or .txt?

**Comment by acutetech** [comment]
All good questions Victor.

A very quick looks shows something of the EXIF format here:
https://docs.fileformat.com/image/exif/

One thing it talks about is that the format can contain a thumbnail, perhaps 160 x 120. Apparently the thumbnail is stored as a JPEG, and would typically be a colour image.

Thinking about not duplicating work: would a 160 x 120 image be OK for transferring to the app when taking test photos during installation? It might be easier to transfer the image as a JPEG and to ask the app to decompress it. Much smaller data, and save the WW from the decompression operation.

If you saved meta data in EXIF then the WW would have to open every (large image) file to read the meta data. Is that OK?

Looks like the fields might be quite complex. Maybe someone would like to review this and establish the minimum requirements for a WW.

Can the camera metadata (e.g. location, settings) be stored in EXIF? Or are there settings that need to be stored in a different way/place?

Regards - Charles

**Comment by Tobyntobyn** [comment]
I like option #2 for metadata coupled with each image. 

160x120 should be sufficient for testing the deployment of the device. It will be blurry but the user will get enough clarity for the set-up. The only disclaimer is when physically testing the device and we find 160x120 to be too blurry. The only other option is to implement the decompression method, which would take some time to solve. 

For the general metadata, how did it work on the 1.0 device for storing it on the SD card, were there any problems that came up while trying to access the metadata? 
I'd suggest the general metadata be stored on the SD card, it enables users to easily access the data if they need it. 
Building on that, the format will depend on what/where the metadata gets used for. 
csv is more approachable for user-readable information (simply opening it in Excel) whereas .json is cleaner to work with within the code. 
I don't suggest we store metadata in .txt format.

One potential problem we might encounter while using EXIF is accessing it within the code. Python would be the desired programming language to import and use the information (eg. displaying GPS information on a map). There isn't an elegant solution in C to achieve this. 
Where exactly in the hierarchy is the step to extract the metadata and display relevant information to the user?

**Comment by victor-wildlife** [comment]
Charles:
Hi Tobyn and Victor

It makes sense to use EXIF for metadata I think. From what I can determine EXIF can be quite big, but need not be. It can include thumbnail images and custom data.

1    Start by assuming we can get the AI board to take photos and save them as jpeg files, on its SD card, with embedded EXIF data including a a thumbnail image....

2    Leave aside for the moment how the files are stored, named etc. Assume a directory listing is somehow available to whichever software needs it...

3    Then the mobile phone app could get a list of files on the Seeed device (or request a one-off test image), then:

4    Ask for the EXIF data from that image. This is a request for data from the app to the Seeed processor, via the WW130 which does not have to understand the request or the response.

5    The  Seeed device provides the app the EXIF data, as a chunk of binary data.

6    The app can do stuff with this data, including extracting the thumbnail and showing it to the operator. The app software environment should have access to plenty of tools and libraries and it makes sense that as much work as possible is done there and as little as possible on the embedded devices.

7    Aside from what the app does, as the Seeed board grabs images it could inform the WW130. The WW130 could request the EXIF data - either all of it or selected fields - and send that via LoRaWAN network to the server. Or it might be easier for the Seeed board to supply the metadata directly rather providing the EXIF data for the WW130 to parse.

There are plenty of tools for playing with EXIF data. One seems to be this: https://exiftool.org

I asked ChatGPT about EXIF libraries. It suggested a library called TinyEXIF, but that seems to be to extract EXIF data: https://github.com/cdcseacave/TinyEXIF

So I asked about code for creating EXIF, and it provided an example. Further questions led to further C programs. I attach these. Of course they are untested, but perhaps this provides a useful kickoff point. My final request was for a program that provided EXIF, including the thumbnail, plus the full-sized image, as a single jpeg file.

As you can see it should not be difficult to add EXIF on the Seeed board and you might like the challenge of trying that.

For my part, I will work on getting the Seeed board to provide EXIF data to the app (more generally, to pass arbitrary-length binary data).

What do you think?

What EXIF tags would we like in our images?

Regards - Charles

**Comment by victor-wildlife** [comment]
I agree 160x120 should be enough to preview the "field of view" of the wildlife watcher on the field while deploying it but I am not sure if thumbnails to the videos/photos are a must though...

We should identify the potential **user cases**. 

I believe the metadata will be extracted and used by:
### User case #1
Operator on the field who would like to preview the field of view of the Wildlife Watcher via the mobile app, 

### User case #2
Biologist who have collected the SD card from the Wildlife Watchers and would like to classify the footage and run ecological analyses 


### User case #3
Biologist who would like to get an alert of an event of interest (e.g. rat detected) via lorawan/4g.....

Am I missing any other user cases?


For user case #1 the flow could be:
1. The mobile phone app requests a one-off test image from the WW130,
2. the WW130 passes the message to the Seed device,
3. the Seed device takes a photo and sends it as a chunk of binary data to the WW130 (it could be 160X120),
4. the WW130 sends the information to the app,
5. the app displays the image to the operator.

This image is temporary so it doesn't require storing much metadata as EXIF


For user case #2 the flow could be:

1. the biologist inserts the sd card in their computer and sees three csv files, a json file and a "media" folder ([example_ww_data here](https://drive.google.com/drive/folders/122PLPlZM5RLRdmdk23KKOzpEoV1ei6cy?usp=drive_link)),
2. uploads the four files and media folder to a long-term storage (e.g. hard drive, AWS, Wildlife Insights...),
3. the camera trap software of choice (e.g. Camtraptor, Wildlife Insights) is able to automatically retrieve all relevant information (e.g. camera serial number, deployment start, deployment end, lat/lon, date/time ....) for the user to analyse/classify the footage. Here is [an R notebook example of how using Camtraptor for the example data](https://colab.research.google.com/drive/1UvX97QVIKlr4_QZ6rA9BYyolQzrEIvKs?usp=sharing)


For user case #3 the flow could be:

1. The Seed device records an image/video and identifies the species of interest,
2. the Seed device saves the image/video on the sd card (as all the other media it recorded),
3. the Seed device send an alert message to the WW130,
4. the WW130 (using lora/G4/WIFI) sends the message to ... TBC (e.g. the db in Azure, trap.NZ, the mobile app, a text message...) 


Let's try to discuss the bigger picture of user cases and once we are on the same page we can narrow down how to keep track of the metadata. @acutetech @Tobyntobyn @CalBou thoughts?
